File: bindings/php/tests/AuthTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

/**
 * crypto_auth - three categories per D-64/D-65: correctness (round trip), rejection (tampered
 * message, wrong key), misuse (wrong-length key/tag - foreclosed at the Rust layer by fixed-size
 * arrays, D-66, so `auth` itself is infallible; only the boundary length checks are testable
 * here).
 */
final class AuthTest extends TestCase
{
    public function testRoundTripsAuthAuthVerify(): void
    {
        $key = dstu_core_auth_keygen();
        $message = 'a message both parties want to confirm is unmodified';
        $tag = dstu_core_auth($key, $message);
        dstu_core_auth_verify($key, $message, $tag);
        $this->expectNotToPerformAssertions();
    }

    public function testRejectsTamperedMessage(): void
    {
        $key = dstu_core_auth_keygen();
        $tag = dstu_core_auth($key, 'original message');
        $this->expectException(DstuCoreException::class);
        dstu_core_auth_verify($key, 'a different message', $tag);
    }

    public function testRejectsWrongKey(): void
    {
        $key = dstu_core_auth_keygen();
        $otherKey = dstu_core_auth_keygen();
        $tag = dstu_core_auth($key, 'message');
        $this->expectException(DstuCoreException::class);
        dstu_core_auth_verify($otherKey, 'message', $tag);
    }

    public function testRejectsWrongLengthKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_auth('too short', 'message');
    }

    public function testRejectsWrongLengthTag(): void
    {
        $key = dstu_core_auth_keygen();
        $this->expectException(\ValueError::class);
        dstu_core_auth_verify($key, 'message', 'too short');
    }
}

File: bindings/go/dstu/auth_test.go
package dstu

import (
	"bytes"
	"testing"
)

// Auth (crypto_auth) - correctness (round trip), rejection (tampered message, wrong key), misuse
// (wrong-length key/tag).

func TestAuthVerifyRoundTrips(t *testing.T) {
	key, err := GenerateAuthKey()
	if err != nil {
		t.Fatal(err)
	}
	defer key.Close()
	message := []byte("a message both parties want to confirm is unmodified")
	tag := key.Compute(message)
	if err := key.Verify(message, tag); err != nil {
		t.Fatal(err)
	}
}

func TestAuthTamperedMessageIsRejected(t *testing.T) {
	key, err := GenerateAuthKey()
	if err != nil {
		t.Fatal(err)
	}
	defer key.Close()
	tag := key.Compute([]byte("original message"))
	if err := key.Verify([]byte("a different message"), tag); err == nil {
		t.Fatal("expected an error")
	}
}

func TestAuthWrongKeyIsRejected(t *testing.T) {
	key, _ := GenerateAuthKey()
	defer key.Close()
	otherKey, _ := GenerateAuthKey()
	defer otherKey.Close()
	message := []byte("message")
	tag := key.Compute(message)
	if err := otherKey.Verify(message, tag); err == nil {
		t.Fatal("expected an error")
	}
}

func TestAuthWrongLengthKeyIsRejected(t *testing.T) {
	if _, err := AuthKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestAuthWrongLengthTagIsRejected(t *testing.T) {
	key, _ := GenerateAuthKey()
	defer key.Close()
	if err := key.Verify([]byte("message"), make([]byte, 4)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestAuthBytesRoundTrip(t *testing.T) {
	key, _ := GenerateAuthKey()
	defer key.Close()
	restored, err := AuthKeyFromBytes(key.Bytes())
	if err != nil {
		t.Fatal(err)
	}
	defer restored.Close()
	if !bytes.Equal(restored.Bytes(), key.Bytes()) {
		t.Fatal("round-tripped key bytes differ")
	}
}
